Implement `ChannelListItem` for new chat design

jai-x opened the pull request:
Part of https://github.com/ppy/osu/issues/17001

Reference design: https://www.figma.com/file/f8b2dHp9LJCMOqYP4mdrPZ/Client%2FChat?node-id=1%3A297

<details>
<summary>Pre-review video</summary>

https://user-images.githubusercontent.com/8362491/158243136-57a6141b-068f-4097-b499-ab6acbd006e1.mp4
</details>

https://user-images.githubusercontent.com/8362491/158266502-8a0a37b7-76bf-47f9-9bb9-93ada90fb088.mp4

Adds new component `ChannelListItem` and it's child components to be
used as the clickable control in the new chat sidebar for joined
channels.

Has public bindables `Unread` and `Mentions` to control the
display of the channel having unread messages or mentions of the user.

Channel select/join requests are exposed via `OnRequestSelect` and
`OnRequestLeave` events respectively which should be handled by a parent
component.

Requires a cached `Bindable<Channel>` instance to be managed by a parent
component.

Requires a cached `OveralayColourScheme` instance to be provided by a
parent component.

### Remarks

~~Component `ControlItemClose` is cribbed from `TabCloseButton` in the current
channel tab control. I'm not quite sure if it looks right but the design did not
specify a way to leave a channel from this control so I added this.~~
Refactored to inherit from `OsuAnimatedButton` and mentioned by Flyte
that a close button should be there on the design.

The mention number has been moved to be right aligned, partly because it
matches design in other familiar chat programs such as Discord, and partly
because I was having trouble figuring out to correctly autosize with grid
container.

~~Overall I'm not too sure on having the `HasUnread` property be externally
managed, but I couldn't think of a clean way for the component to bind to
the channel events and correctly handle auto changing between read and
unread states.~~
Changed to use bindable flow, from PR review.

jai-x committed:
Implement `ChannelControlItem` for new chat design

Adds new component `ChannelControlItem` and it's child components to be
used as the clickable control in the new chat sidebar for joined
channels.

Has public properties `HasUnread` and `MentionCount` to control the
display of the channel having unread messages or mentions of the user.

Channel select/join requests are exposed via `OnRequestSelect` and
`OnRequestLeave` events respectively which should be handled by a parent
component.

Requires a cached `Bindable<Channel>` instance to be managed by a parent
component.

Requires a cached `OveralayColourScheme` instance to be provided by a
parent component.

bdach commented on osu.Game.Tests/Visual/Online/TestSceneChannelControlItem.cs:
```diff
@@ -0,0 +1,145 @@
+                    {
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                        Y = -140,
```
It's test code so I won't nitpick too hard but using negative Y like this is in general a weird way to layout. I'd just use a vertical fill flow.

bdach commented on osu.Game.Tests/Visual/Online/TestSceneChannelControlItem.cs:
```diff
@@ -0,0 +1,145 @@
+                    {
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                        Size = new Vector2(190, 200),
```
Similarly, really weird how this has hardcoded height instead of autosizing to children.

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs:
```diff
@@ -0,0 +1,167 @@
+        public event Action<Channel>? OnRequestSelect;
+        public event Action<Channel>? OnRequestLeave;
+
+        public int MentionCount
+        {
+            get => mention?.MentionCount ?? 0;
+            set
+            {
+                if (mention != null)
+                    mention.MentionCount = value;
+            }
+        }
```
These sorts of properties come out way cleaner using bindable flow. Less worrying about whether the underlying thing being delegated to exists etc.

jai-x replied:
Good point. For some reason using bindable flow for this slipped my mind even though I was using it for the selected channel state. Will change.

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs:
```diff
@@ -0,0 +1,167 @@
+        private ControlItemMention? mention;
+        private ControlItemClose? close;
+
+        [Resolved]
+        private Bindable<Channel>? selectedChannel { get; set; }
+
+        [Resolved]
+        private OverlayColourProvider? colourProvider { get; set; }
```
Why are these marked nullable if they're not `CanBeNull = true`? It's not idiomatic, if a dependency is expected to be injected at BDL time then we usually make it not nullable and then initialise it with `null!` to make the compiler shut up.

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs:
```diff
@@ -0,0 +1,167 @@
+        [Resolved]
+        private OverlayColourProvider? colourProvider { get; set; }
+
+        private readonly Channel channel;
```
Readonly field should be first

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs:
```diff
@@ -0,0 +1,167 @@
+                hoverBox = new Box
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = colourProvider!.Background3,
```
We generally refrain from using null-forgiving operators unless there is absolutely no alternative considered better. Which there is, in this case - see comment wrt BDL-injected field nullability above.

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs:
```diff
@@ -0,0 +1,167 @@
+                new Container
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Padding = new MarginPadding { Left = 18, Right = 5 },
```
Not sure this padding should be allowed to be asymmetric like this, it looks pretty bad to me. Would make it symmetric.

jai-x replied:
Given the reference design doesn't show any channel text reaching the extent of the right padding, I took cues from Discord's design for this asymmetrical padding.
I think having it symmetrical will make the text look too cluttered, especially since the mention number and the close button push the text in from the right.
I agree it looks too close on the right side though, so I've made that side larger a bit.

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs:
```diff
@@ -0,0 +1,167 @@
+            }, true);
+
+            Action = () => OnRequestSelect?.Invoke(channel);
+            close!.Action = () => OnRequestLeave?.Invoke(channel);
```
Going to be interested in other opinions on this because I have had the same problem - this sort of assignment with null-forgiveness is what initialising drawables in BDL with nullability enabled inevitably leads to, and to me this reads really badly.

frenzibyte replied:
In this specific case, the actiosn could be assigned on object initializer rather than here, i.e.:
```diff
diff --git a/osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs b/osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs
index e475ac7821..4d989ebdfd 100644
--- a/osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs
+++ b/osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs
@@ -114,12 +114,15 @@ private void load()
                                     Anchor = Anchor.CentreLeft,
                                     Origin = Anchor.CentreLeft,
                                     Margin = new MarginPadding { Right = 3 },
+                                    Action = () => OnRequestLeave?.Invoke(channel),
                                 }
                             }
                         },
                     },
                 },
             };
+
+            Action = () => OnRequestSelect?.Invoke(channel);
         }

         protected override void LoadComplete()
@@ -133,9 +136,6 @@ protected override void LoadComplete()
                 else
                     selectBox?.Hide();
             }, true);
-
-            Action = () => OnRequestSelect?.Invoke(channel);
-            close!.Action = () => OnRequestLeave?.Invoke(channel);
         }

         protected override bool OnHover(HoverEvent e)
```

But in any other case, I think it's safe to say those BDL fields will be non-null in any other place in the component itself, so they can be declared as non-nullabe with `= null!` appended at the end.

bdach replied:
You'd say that [but I've been told differently otherwise in the past](https://github.com/ppy/osu/pull/16264#discussion_r775506750). Also see [this thread](https://github.com/ppy/osu/pull/17014#discussion_r816237281).

frenzibyte replied:
Interesting opinions, I can rarely think of any local `Drawable` component that has logic executed before BDL myself, so thought using `= null!` makes sense. I guess `AsNonNull()` is the way then...? I'm not sure.

peppy replied:
We've started to do the same in realm as well. Difference here is that realm guarantees to populate in the ctor, while here we lose null protection for things like properties which may access a field before BDL has been run.

bdach replied:
I'd still like to see these particular assignments of `Action` to be moved to BDL as proposed above.

peppy replied:
We've come to the general conclusion of using `=null!` suppression for BDL usage, so you're welcome to apply that here if you'd like to.

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItemAvatar.cs:
```diff
@@ -0,0 +1,60 @@
+
+            Children = new Drawable[]
+            {
+                new SpriteIcon
+                {
+                    Icon = FontAwesome.Solid.At,
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    Colour = Colour4.Black,
+                    RelativeSizeAxes = Axes.Both,
+                    Alpha = 0.2f,
+                },
```
This placeholder "at" icon is a really weird detail to add given that if I hadn't read source I wouldn't have even noticed it was there. It may be a good idea to have it but it really ought to be more visible than that imo.

jai-x replied:
This is following the design of the existing placeholder that user avatars have in the current channel tab control.
It looked good enough to keep but I can remove it if it seems superfluous.

bdach replied:
I'm not saying that it needs to go entirely, just that a 0.2 opacity black "at" sign on top of an already-dark background is doing way less here than what it was doing on the old chat design. Context is important.

jai-x replied:
Adjusted to be faded from white and will hide on avatar load.

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItemClose.cs:
```diff
@@ -0,0 +1,55 @@
+
+        protected override bool OnHover(HoverEvent e)
+        {
+            icon.FadeColour(Color4.Red, 200, Easing.OutQuint);
```
This red doesn't look great, too saturated in comparison with everything else. Would recommend trying `OsuColour.Red1` or similar.

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItemMention.cs:
```diff
@@ -0,0 +1,77 @@
+                new Box
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = colourProvider!.Colour1,
```
Not sure why this is in the colour provider theme rather than yellow as on design. There's really not enough contrast right now to be able to easily see that there's a channel with mentions, which a bright colour with an entirely distinct hue would help with.

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItemText.cs:
```diff
@@ -0,0 +1,71 @@
+                return;
+
+            if (HasUnread)
+                text!.Colour = Colour4.White;
```
`colourProvider.Content1` would be better

jai-x committed:
Code quality fixes

jai-x committed:
Use autosizing for ChannelControlItem test scene

jai-x committed:
Mark required dependencies as non-nullable

jai-x committed:
Adjust ControlItemAvatar placeholder animation and colour

jai-x committed:
Use OsuColour.Red1 for ControlItemClose hover colour

jai-x committed:
Use OsuColour.YellowLight for ControlItemMention background

jai-x committed:
Use ColourProvider.Content1 for ControlItemText colour

jai-x committed:
Refactor ControlItemMention to use bindable flow

jai-x committed:
Refactor ControlItemText to use bindable flow for unread state

jai-x committed:
Adjust ControlItem padding

jai-x committed:
Code quality fixes

jai-x committed:
Ensure existing items are expired and cleared in ChannelControlItem test setup

peppy commented:
What's with the huge vertical spacing between each item?

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs:
```diff
@@ -0,0 +1,151 @@
+        public event Action<Channel>? OnRequestSelect;
+        public event Action<Channel>? OnRequestLeave;
+
+        [Cached]
+        public readonly BindableInt Mentions = new BindableInt();
+
+        [Cached]
+        public readonly BindableBool Unread = new BindableBool();
```
Caching dependencies to direct children is generally frowned upon. Preferred style is to either pass down directly for everything that isn't a bindable, and for bindables, it's to declare a matching bindable on the child components and then use `BindTarget` to set up a binding. Examples of this are all over, one is [here](https://github.com/ppy/osu/blob/f81e5e8b992d73411277593b32697d3fc80cf7be/osu.Game/Beatmaps/Drawables/Cards/BeatmapCardNormal.cs#L85).

bdach commented on osu.Game/Overlays/Chat/ChannelControl/ControlItemMention.cs:
```diff
@@ -0,0 +1,72 @@
+                new Box
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = osuColour.YellowLight,
```
```suggestion
                    Colour = osuColour.Orange1,
```

is what is actually spec'd on design:

![2022-03-15-223747_726x241_scrot](https://user-images.githubusercontent.com/20418176/158477172-283e4aa6-1868-4afe-a7ed-27efa5259489.png)

Check out the "basic" colour theme here: https://www.figma.com/file/VIkXMYNPMtQem2RJg9k2iQ/Asset%2FColours?node-id=1838%3A3 if you haven't seen that already.

jai-x committed:
Move `ControlItem` Action assignments into BDL

jai-x committed:
Use `Orange1` for `ControlItemMention` background colour

jai-x committed:
Use `BindTarget` instead of caching for `ControlItem` mentions bindable flow

jai-x committed:
Use `BindTarget` instead of caching for `ControlItem` unread flow

peppy commented:
> I just added spacing to have the items clearly separated just for visual test purposes. No functional reason.

Ah, I see. Didn't realise the list wasn't part of what was being tested. I'd probably make the spacing match the design so it's easier to visually compare during review.

jai-x committed:
Remove redundant spacing in `ControlItem` test scene

bdach committed:
Remove unnecessary `OsuColour` cache

One is already cached at `OsuGameBase` level.

bdach committed:
Simplify fade-out transform

bdach committed:
Move dependencies out of fields to BDL args where possible

bdach commented:
I've pushed in some more minor cleanups, please check when you get the chance.

Other than that source looks sane to me.

peppy committed:
Add missing `nullable`

peppy committed:
Fix incorrect usage of nullable in `ControlItemMention`

peppy commented:
Some of the flow is very uncertain with this stand-alone PR, so I'll just smile and nod until a future PR that hooks things up (ie. the unread state, the mention count not disappearing when channel is selected etc.)

peppy commented on osu.Game/Overlays/Chat/ChannelControl/ControlItemText.cs:
```diff
@@ -0,0 +1,59 @@
+
+namespace osu.Game.Overlays.Chat.ChannelControl
+{
+    public class ControlItemText : Container
```
Would also argue there's no point in adding this class. It's 99% data binding from the parent just to update one string.

peppy committed:
Rename icon creation method

peppy commented on osu.Game/Overlays/Chat/ChannelControl/ControlItemClose.cs:
```diff
@@ -0,0 +1,59 @@
+            };
+        }
+
+        protected override bool OnMouseDown(MouseDownEvent e)
+        {
+            icon.ScaleTo(0.5f, 1000, Easing.OutQuint);
+            return base.OnMouseDown(e);
+        }
+
+        protected override void OnMouseUp(MouseUpEvent e)
+        {
+            icon.ScaleTo(0.75f, 1000, Easing.OutElastic);
+            base.OnMouseUp(e);
+        }
+
```
This should probably be using the newer transform logic (see https://github.com/ppy/osu/blob/18c144395ae7644601b09da7a112d963a0bcb27c/osu.Game/Graphics/UserInterface/OsuAnimatedButton.cs#L98-L126). The scale is too harsh (looks to have been copied from the old tab implementation).

peppy commented on osu.Game/Overlays/Chat/ChannelControl/ControlItem.cs:
```diff
@@ -0,0 +1,148 @@
+        [Resolved]
+        private Bindable<Channel> selectedChannel { get; set; } = null!;
+
+        public ControlItem(Channel channel)
```
Rename to `ChannelTab` or `ChannelListItem`, along with all the related classes.

peppy commented on osu.Game/Overlays/Chat/ChannelControl/ControlItemAvatar.cs:
```diff
@@ -0,0 +1,62 @@
+                    RelativeSizeAxes = Axes.Both,
+                    Alpha = 0.5f,
+                },
+                new DelayedLoadWrapper(avatar = new DrawableAvatar(channel.Users.First())
```
Did you consider using `UpdateableAvatar` which has built-in `DelayedLoad` and placeholder logic?

peppy reviewed (changes requested):
I'll do a final animation pass once the class naming has been resolved.

jai-x committed:
Remove redundant `ControlItemText` class

jai-x committed:
Use `OsuAnimatedButton` as base for `ControlItemClose`

jai-x committed:
Use `UpdateableAvatar` and remove redundant `ControlItemAvatar` class

jai-x committed:
Rename `ControlItem` -> `ChannelListItem` and associated classes

jai-x commented:
> Some of the flow is very uncertain with this stand-alone PR, so I'll just smile and nod until a future PR that hooks things up (ie. the unread state, the mention count not disappearing when channel is selected etc.)

My reasoning for this is that the overall state for the chat (channel reads, join/leave etc) is managed in the parent overlay component where it interacts with the channel manager and should probably be kept centralised there in the new overlay where it can dispatch state to the child component such as this one.

Additionally, having the item manually clear the read/mention state may not be intended in cases where the user clicks on an item and then clicks to another before the `DrawableChannel` is loaded and the channel gets marked as being read before it was even loaded and shown to the user.

jai-x committed:
Pass correct params to `UpdateableAvatar` ctor in `ChannelListItem`

peppy committed:
Merge branch 'master' into new-chat-channel-control

peppy committed:
Rename renaming button/pill classes

peppy commented:
Not sure about the box background around the close button (I think you may have misinterpreted what I meant regarding the use of `OsuAnimatedButton`, but I did have a feeling my wording wasn't the best). I'll change this back to how you had it but use the more correct scale values.

peppy committed:
Remove box surrounding close button

peppy committed:
Add slight fade when hovering `ChannelListItem`s

peppy committed:
Add animation when mention pill appears (or a new message arrives)

peppy committed:
Add step to `TestSceneChannelListItem` to actually select an item

peppy committed:
Add slight animation when highlighting an unread channel

peppy committed:
Remove space between '#' and channel name

This was in flyte's design (he was following discord) but doesn't feel
too correct when it's not aligned with anything else. We can revisit
this if/when we have better glyphs to represent channel types.

peppy committed:
Also add fade to `ChannelItem` selected box

peppy committed:
Fix close button animation not being applied correct on mouse down due to conflicting scales
